File: src/NexusUnleashed.Sts/AuthFlow.cs
using System;
using System.Security.Cryptography;
using System.Threading.Tasks;
using NexusUnleashed.Cryptography;

namespace NexusUnleashed.Sts;

public interface IAccountStore
{
    Task<(byte[] Salt, byte[] Verifier)?> GetSrpCredentialsAsync(string loginName);
    Task StoreGameTokenAsync(string loginName, Guid token);
    Task<long> GetUserIdAsync(string loginName);
}

public static class AuthFlow
{
    private const string KeyLogin = "auth.login";
    private const string KeySrp = "auth.srp";
    private const string KeyAuthed = "auth.ok";
    private const string KeySession = "auth.sessionkey";

    public static void Register(StsServer server, IAccountStore accounts)
    {
        server.On("/Sts/Connect", (s, r) => s.SendAsync(StsReply.Ok(r.Sequence, "")));
        server.On("/Sts/Ping", (s, r) => s.SendAsync(StsReply.Ok(r.Sequence, "")));

        server.On("/Auth/LoginStart", async (s, r) =>
        {
            string login = XmlBody.Field(r.BodyText, "LoginName");
            s.State[KeyLogin] = login;

            var creds = await accounts.GetSrpCredentialsAsync(login);
            if (creds == null || creds.Value.Verifier.Length == 0)
            {
                await s.SendAsync(StsReply.Error(r.Sequence, 403));                return;
            }

            var srp = new StsSrp(creds.Value.Salt, creds.Value.Verifier, login);
            byte[] B = srp.StartHandshake();            s.State[KeySrp] = srp;
            Console.WriteLine("[STS-SRP] LoginStart: game-SRP (little-endian)");

            byte[] blob = KeyDataBlob.Pack(creds.Value.Salt, B);
            await s.SendAsync(StsReply.Ok(r.Sequence, KeyDataBody(blob)));
        });

        server.On("/Auth/KeyData", async (s, r) =>
        {
            if (s.State[KeySrp] is not StsSrp srp)
            {
                await s.SendAsync(StsReply.Error(r.Sequence, 409));                return;
            }

            byte[] clientBlob = Convert.FromBase64String(XmlBody.Field(r.BodyText, "KeyData"));
            var parts = KeyDataBlob.Unpack(clientBlob);
            byte[] a = parts.Length > 0 ? parts[0] : Array.Empty<byte>();
            byte[] m1 = parts.Length > 1 ? parts[1] : Array.Empty<byte>();

            if (!srp.Verify(a, m1, out byte[] m2, out byte[] sessionKey))
            {
                Console.WriteLine($"[STS-SRP] KeyData proof did NOT verify (A={a.Length}B, M1={Convert.ToHexString(m1)})");
                await s.SendAsync(StsReply.Error(r.Sequence, 403));                return;
            }
            Console.WriteLine($"[STS-SRP] proof VERIFIED — variant: {srp.MatchedVariant}");
            Console.WriteLine($"[STS-SRP] K={Convert.ToHexString(sessionKey)}");
            s.State[KeyAuthed] = true;
            s.State[KeySession] = sessionKey;

            byte[] blob = KeyDataBlob.Pack(m2);
            await s.SendAsync(StsReply.Ok(r.Sequence, KeyDataBody(blob)));
            s.EnableEncryption(sessionKey);
        });

        server.On("/Auth/LoginFinish", async (s, r) =>
        {
            string login = s.State.TryGetValue(KeyLogin, out var lo) ? (string)lo : "";
            long uid = await accounts.GetUserIdAsync(login);
            string body =
                "<Reply>\n" +
                "<AuthType>Password</AuthType>\n" +
                $"<UserId>{uid}</UserId>\n" +
                "<UserCenter>1</UserCenter>\n" +
                $"<UserName>{System.Security.SecurityElement.Escape(login)}</UserName>\n" +
                "<LocationId>1</LocationId>\n" +
                "<AccessMask>4294967295</AccessMask>\n" +
                "<Status>0</Status>\n" +
                "<Roles>1</Roles>\n" +
                "</Reply>\n";
            await s.SendAsync(StsReply.Ok(r.Sequence, body));
        });

        server.On("/GameAccount/ListMyAccounts", async (s, r) =>
        {
            Console.WriteLine($"[STS] ListMyAccounts req: {r.BodyText.Replace("\n", " ")}");
            string login = s.State.TryGetValue(KeyLogin, out var lo) ? (string)lo : "";
            long uid = await accounts.GetUserIdAsync(login);
            string e = System.Security.SecurityElement.Escape(login) ?? "";
            string alias = login.Contains('@') ? login[..login.IndexOf('@')] : login;
            string ea = System.Security.SecurityElement.Escape(alias) ?? "";
            string body =
                "<Reply>\n<GameAccount>\n" +
                $"<GameAccountId>{uid}</GameAccountId>\n" +
                $"<AccountId>{uid}</AccountId>\n" +
                $"<LoginName>{e}</LoginName>\n" +
                $"<UserId>{uid}</UserId>\n" +
                $"<UserName>{e}</UserName>\n" +
                $"<Email>{e}</Email>\n" +
                $"<Alias>{ea}</Alias>\n" +
                $"<AccountAlias>{ea}</AccountAlias>\n" +
                "<GameCode>wildstar</GameCode>\n" +
                "<AppId>0</AppId>\n" +
                "<UserCenter>1</UserCenter>\n" +
                "<State>1</State>\n" +
                "<Status>0</Status>\n" +
                "<Roles>1</Roles>\n" +
                "</GameAccount>\n</Reply>\n";
            await s.SendAsync(StsReply.Ok(r.Sequence, body));
        });

        server.On("/Auth/RequestGameToken", async (s, r) =>
        {
            if (s.State.TryGetValue(KeyAuthed, out var ok) && ok is true &&
                s.State.TryGetValue(KeyLogin, out var login))
            {
                Guid token = NewToken();
                await accounts.StoreGameTokenAsync((string)login, token);
                long uid = await accounts.GetUserIdAsync((string)login);
                AuthSession.Register(token.ToString("N"), uid);
                await s.SendAsync(StsReply.Ok(r.Sequence,
                    XmlBody.Fields(("Token", token.ToString("N")))));
            }
            else
            {
                await s.SendAsync(StsReply.Error(r.Sequence, 401));            }
        });
    }

    private static string KeyDataBody(byte[] blob)
        => "<Reply>\n<KeyData>" + Convert.ToBase64String(blob) + "</KeyData>\n</Reply>\n";

    private static Guid NewToken()
    {
        Span<byte> b = stackalloc byte[16];
        RandomNumberGenerator.Fill(b);
        return new Guid(b);
    }
}

internal static class XmlBody
{
    public static string Fields(params (string Tag, string Value)[] fields)
    {
        var sb = new System.Text.StringBuilder("<Reply>\n");
        foreach (var (tag, val) in fields)
            sb.Append('<').Append(tag).Append('>')
              .Append(System.Security.SecurityElement.Escape(val))
              .Append("</").Append(tag).Append(">\n");
        return sb.Append("</Reply>\n").ToString();
    }

    public static string Field(string xml, string tag)
    {
        string open = "<" + tag + ">", close = "</" + tag + ">";
        int i = xml.IndexOf(open, StringComparison.Ordinal);
        if (i < 0) return "";
        i += open.Length;
        int j = xml.IndexOf(close, i, StringComparison.Ordinal);
        return j < 0 ? "" : xml[i..j];
    }

    public static string? FirstText(string xml)
    {
        int gt = xml.IndexOf('>');
        if (gt < 0) return xml.Trim().Length > 0 ? xml.Trim() : null;
        int lt = xml.IndexOf('<', gt + 1);
        if (lt < 0) return null;
        string inner = xml[(gt + 1)..lt].Trim();
        return inner.Length > 0 ? inner : null;
    }
}

internal static class KeyDataBlob
{
    public static byte[] Pack(params byte[][] parts)
    {
        int n = 0;
        foreach (var p in parts) n += 4 + p.Length;
        var buf = new byte[n];
        int o = 0;
        foreach (var p in parts)
        {
            buf[o] = (byte)p.Length; buf[o + 1] = (byte)(p.Length >> 8);
            buf[o + 2] = (byte)(p.Length >> 16); buf[o + 3] = (byte)(p.Length >> 24);
            System.Array.Copy(p, 0, buf, o + 4, p.Length);
            o += 4 + p.Length;
        }
        return buf;
    }

    public static byte[][] Unpack(byte[] blob)
    {
        var list = new System.Collections.Generic.List<byte[]>();
        int o = 0;
        while (o + 4 <= blob.Length)
        {
            int len = blob[o] | (blob[o + 1] << 8) | (blob[o + 2] << 16) | (blob[o + 3] << 24);
            o += 4;
            if (len < 0 || o + len > blob.Length) break;
            var p = new byte[len];
            System.Array.Copy(blob, o, p, 0, len);
            list.Add(p);
            o += len;
        }
        return list.ToArray();
    }
}

internal static class Hex
{
    public static string To(byte[] b)
    {
        var c = new char[b.Length * 2];
        const string h = "0123456789abcdef";
        for (int i = 0; i < b.Length; i++) { c[i * 2] = h[b[i] >> 4]; c[i * 2 + 1] = h[b[i] & 0xF]; }
        return new string(c);
    }

    public static byte[] From(string hex)
    {
        if (string.IsNullOrEmpty(hex) || (hex.Length & 1) == 1) return Array.Empty<byte>();
        var b = new byte[hex.Length / 2];
        for (int i = 0; i < b.Length; i++)
            b[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        return b;
    }
}

File: cpp/src/sts/auth_flow.cpp
#include "sts/auth_flow.h"
#include "crypto/sts_srp.h"
#include <openssl/rand.h>
#include <atomic>
#include <mutex>
#include <unordered_map>
#include <string>

namespace nexus::sts {

using Bytes = std::vector<uint8_t>;

namespace {
    std::mutex g_mtx;
    std::unordered_map<std::string, long> g_token_to_account;
    std::atomic<long> g_last_account{0};
}
void AuthSession::Register(const std::string& tokenHex, long accountId) {
    { std::lock_guard<std::mutex> lk(g_mtx); if (!tokenHex.empty()) g_token_to_account[tokenHex] = accountId; }
    g_last_account = accountId;
}
long AuthSession::ResolveToken(const std::string& tokenHex) {
    std::lock_guard<std::mutex> lk(g_mtx);
    auto it = g_token_to_account.find(tokenHex);
    return it == g_token_to_account.end() ? 0 : it->second;
}
long AuthSession::LastAccountId() { return g_last_account.load(); }

static const char* B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static std::string base64_encode(const Bytes& in) {
    std::string out;
    size_t i = 0;
    while (i + 3 <= in.size()) {
        uint32_t n = (in[i] << 16) | (in[i+1] << 8) | in[i+2];
        out += B64[(n >> 18) & 63]; out += B64[(n >> 12) & 63];
        out += B64[(n >> 6) & 63];  out += B64[n & 63];
        i += 3;
    }
    size_t rem = in.size() - i;
    if (rem == 1) {
        uint32_t n = in[i] << 16;
        out += B64[(n >> 18) & 63]; out += B64[(n >> 12) & 63]; out += "==";
    } else if (rem == 2) {
        uint32_t n = (in[i] << 16) | (in[i+1] << 8);
        out += B64[(n >> 18) & 63]; out += B64[(n >> 12) & 63]; out += B64[(n >> 6) & 63]; out += '=';
    }
    return out;
}
static int b64val(char c) {
    if (c >= 'A' && c <= 'Z') return c - 'A';
    if (c >= 'a' && c <= 'z') return c - 'a' + 26;
    if (c >= '0' && c <= '9') return c - '0' + 52;
    if (c == '+') return 62;
    if (c == '/') return 63;
    return -1;
}
static Bytes base64_decode(const std::string& s) {
    Bytes out; int buf = 0, bits = 0;
    for (char c : s) {
        if (c == '=' || c == '\r' || c == '\n' || c == ' ') continue;
        int v = b64val(c); if (v < 0) continue;
        buf = (buf << 6) | v; bits += 6;
        if (bits >= 8) { bits -= 8; out.push_back(static_cast<uint8_t>((buf >> bits) & 0xFF)); }
    }
    return out;
}

static std::string xml_escape(const std::string& s) {
    std::string o; o.reserve(s.size());
    for (char c : s) switch (c) {
        case '&': o += "&amp;"; break;  case '<': o += "&lt;"; break;
        case '>': o += "&gt;"; break;   case '"': o += "&quot;"; break;
        case '\'': o += "&apos;"; break; default: o += c;
    }
    return o;
}
static std::string xml_field(const std::string& xml, const std::string& tag) {
    std::string open = "<" + tag + ">", close = "</" + tag + ">";
    size_t i = xml.find(open); if (i == std::string::npos) return "";
    i += open.size();
    size_t j = xml.find(close, i);
    return j == std::string::npos ? "" : xml.substr(i, j - i);
}

static Bytes blob_pack(std::initializer_list<Bytes> parts) {
    Bytes buf;
    for (const auto& p : parts) {
        uint32_t n = static_cast<uint32_t>(p.size());
        buf.push_back(uint8_t(n)); buf.push_back(uint8_t(n >> 8));
        buf.push_back(uint8_t(n >> 16)); buf.push_back(uint8_t(n >> 24));
        buf.insert(buf.end(), p.begin(), p.end());
    }
    return buf;
}
static std::vector<Bytes> blob_unpack(const Bytes& blob) {
    std::vector<Bytes> list; size_t o = 0;
    while (o + 4 <= blob.size()) {
        uint32_t len = blob[o] | (blob[o+1] << 8) | (blob[o+2] << 16) | (uint32_t(blob[o+3]) << 24);
        o += 4;
        if (o + len > blob.size()) break;
        list.emplace_back(blob.begin() + o, blob.begin() + o + len);
        o += len;
    }
    return list;
}

static std::string new_token_hex() {
    uint8_t b[16];
    RAND_bytes(b, 16);
    static const char* h = "0123456789abcdef";
    std::string s; s.reserve(32);
    for (int i = 0; i < 16; ++i) { s += h[b[i] >> 4]; s += h[b[i] & 0xF]; }
    return s;
}

static std::string key_data_body(const Bytes& blob) {
    return "<Reply>\n<KeyData>" + base64_encode(blob) + "</KeyData>\n</Reply>\n";
}

void AuthFlow::Register(StsServer& server, IAccountStore& accounts) {
    server.On("/Sts/Connect", [](StsSessionState&, const StsRequest& r) -> HandlerResult {
        return { StsReply::Ok(r.sequence(), ""), std::nullopt };
    });
    server.On("/Sts/Ping", [](StsSessionState&, const StsRequest& r) -> HandlerResult {
        return { StsReply::Ok(r.sequence(), ""), std::nullopt };
    });

    server.On("/Auth/LoginStart", [&accounts](StsSessionState& st, const StsRequest& r) -> HandlerResult {
        st.login = xml_field(r.body_text(), "LoginName");
        auto creds = accounts.GetSrpCredentials(st.login);
        if (!creds || creds->second.empty())
            return { StsReply::Error(r.sequence(), 403), std::nullopt };
        st.srp = std::make_shared<crypto::StsSrp>(creds->first, creds->second, st.login);
        Bytes B = st.srp->StartHandshake();
        std::printf("[STS-SRP] LoginStart: game-SRP (little-endian)\n");
        Bytes blob = blob_pack({ creds->first, B });
        return { StsReply::Ok(r.sequence(), key_data_body(blob)), std::nullopt };
    });

    server.On("/Auth/KeyData", [](StsSessionState& st, const StsRequest& r) -> HandlerResult {
        if (!st.srp)
            return { StsReply::Error(r.sequence(), 409), std::nullopt };
        Bytes clientBlob = base64_decode(xml_field(r.body_text(), "KeyData"));
        auto parts = blob_unpack(clientBlob);
        Bytes a  = parts.size() > 0 ? parts[0] : Bytes{};
        Bytes m1 = parts.size() > 1 ? parts[1] : Bytes{};
        Bytes m2, K;
        if (!st.srp->Verify(a, m1, m2, K)) {
            std::printf("[STS-SRP] KeyData proof did NOT verify (A=%zuB)\n", a.size());
            return { StsReply::Error(r.sequence(), 403), std::nullopt };
        }
        std::printf("[STS-SRP] proof VERIFIED (game-SRP little-endian)\n");
        st.authed = true;
        st.session_key = K;
        Bytes blob = blob_pack({ m2 });
        return { StsReply::Ok(r.sequence(), key_data_body(blob)), K };
    });

    server.On("/Auth/LoginFinish", [&accounts](StsSessionState& st, const StsRequest& r) -> HandlerResult {
        long uid = accounts.GetUserId(st.login);
        std::string e = xml_escape(st.login);
        std::string body =
            "<Reply>\n<AuthType>Password</AuthType>\n<UserId>" + std::to_string(uid) +
            "</UserId>\n<UserCenter>1</UserCenter>\n<UserName>" + e +
            "</UserName>\n<LocationId>1</LocationId>\n<AccessMask>4294967295</AccessMask>\n"
            "<Status>0</Status>\n<Roles>1</Roles>\n</Reply>\n";
        return { StsReply::Ok(r.sequence(), body), std::nullopt };
    });

    server.On("/GameAccount/ListMyAccounts", [&accounts](StsSessionState& st, const StsRequest& r) -> HandlerResult {
        long uid = accounts.GetUserId(st.login);
        std::string e = xml_escape(st.login);
        std::string alias = st.login.find('@') != std::string::npos ? st.login.substr(0, st.login.find('@')) : st.login;
        std::string ea = xml_escape(alias);
        std::string u = std::to_string(uid);
        std::string body =
            "<Reply>\n<GameAccount>\n<GameAccountId>" + u + "</GameAccountId>\n<AccountId>" + u +
            "</AccountId>\n<LoginName>" + e + "</LoginName>\n<UserId>" + u + "</UserId>\n<UserName>" + e +
            "</UserName>\n<Email>" + e + "</Email>\n<Alias>" + ea + "</Alias>\n<AccountAlias>" + ea +
            "</AccountAlias>\n<GameCode>wildstar</GameCode>\n<AppId>0</AppId>\n<UserCenter>1</UserCenter>\n"
            "<State>1</State>\n<Status>0</Status>\n<Roles>1</Roles>\n</GameAccount>\n</Reply>\n";
        return { StsReply::Ok(r.sequence(), body), std::nullopt };
    });

    server.On("/Auth/RequestGameToken", [&accounts](StsSessionState& st, const StsRequest& r) -> HandlerResult {
        if (st.authed && !st.login.empty()) {
            std::string token = new_token_hex();
            accounts.StoreGameToken(st.login, token);
            long uid = accounts.GetUserId(st.login);
            AuthSession::Register(token, uid);
            std::string body = "<Reply>\n<Token>" + token + "</Token>\n</Reply>\n";
            return { StsReply::Ok(r.sequence(), body), std::nullopt };
        }
        return { StsReply::Error(r.sequence(), 401), std::nullopt };
    });
}

} // namespace nexus::sts
